bgpd/bgp_pbr.c

```diff
@@ -459,6 +504,373 @@ void bgp_pbr_print_policy_route(struct bgp_pbr_entry_main *api)
 	zlog_info("%s", return_string);
 }
 
+static void bgp_pbr_flush_entry(struct bgp *bgp, struct bgp_pbr_action *bpa,
+				struct bgp_pbr_match *bpm,
+				struct bgp_pbr_match_entry *bpme)
+{
+	/* if bpme is null, bpm is also null
+	 */
+	if (bpme == NULL)
+		return;
+	/* ipset del entry */
+	if (bpme->installed) {
+		bgp_send_pbr_ipset_entry_match(bpme, false);
+		bpme->installed = false;
+		bpme->backpointer = NULL;
+	}
+	hash_release(bpm->entry_hash, bpme);
+	if (hashcount(bpm->entry_hash) == 0) {
+		/* delete iptable entry first */
+		/* then delete ipset match */
+		if (bpm->installed) {
+			if (bpm->installed2) {
+				bgp_send_pbr_iptable(bpm->action,
+						     bpm, false);
+				bpm->installed2 = false;
+			}
+			bgp_send_pbr_ipset_match(bpm, false);
+			bpm->installed = false;
+			bpm->action = NULL;
+		}
+		hash_release(bgp->pbr_match_hash, bpm);
+		/* XXX release pbr_match_action if not used
+		 * note that drop does not need to call send_pbr_action
+		 */
+	}
+}
+
+struct bgp_pbr_match_entry_remain {
+	struct bgp_pbr_match_entry *bpme_to_match;
+	struct bgp_pbr_match_entry *bpme_found;
+};
+
+static int bgp_pbr_get_remaining_entry(struct hash_backet *backet, void *arg)
+{
+	struct bgp_pbr_match *bpm = (struct bgp_pbr_match *)backet->data;
+	struct bgp_pbr_match_entry_remain *bpmer =
+		(struct bgp_pbr_match_entry_remain *)arg;
+	struct bgp_pbr_match *bpm_temp;
+	struct bgp_pbr_match_entry *bpme = bpmer->bpme_to_match;
+
+	if (!bpme->backpointer ||
+	    bpm == bpme->backpointer ||
+	    bpme->backpointer->action == bpm->action)
+		return HASHWALK_CONTINUE;
+	/* ensure bpm other characteristics are equal */
+	bpm_temp = bpme->backpointer;
+	if (bpm_temp->vrf_id != bpm->vrf_id ||
+	    bpm_temp->type != bpm->type ||
+	    bpm_temp->flags != bpm->flags)
+		return HASHWALK_CONTINUE;
+
+	/* look for remaining bpme */
+	bpmer->bpme_found = hash_lookup(bpm->entry_hash, bpme);
+	if (!bpmer->bpme_found)
+		return HASHWALK_CONTINUE;
+	return HASHWALK_ABORT;
+}
+
+static void bgp_pbr_policyroute_remove_from_zebra(struct bgp *bgp,
+						  struct bgp_info *binfo,
+						  vrf_id_t vrf_id,
+						  struct prefix *src,
+						  struct prefix *dst)
+{
+	struct bgp_pbr_match temp;
+	struct bgp_pbr_match_entry temp2;
+	struct bgp_pbr_match *bpm;
+	struct bgp_pbr_match_entry *bpme;
+	struct bgp_pbr_match_entry_remain bpmer;
+
+	/* as we don't know information from EC
+	 * look for bpm that have the bpm
+	 * with vrf_id characteristics
+	 */
+	memset(&temp2, 0, sizeof(temp2));
+	memset(&temp, 0, sizeof(temp));
+	if (src)
```

Comment:
we should really gather up all the src and dst if statements into one.  This seemslike an odd way to do this

Reply:
ok

Reply:
There are multiple `if (src) .... ` `if (dst) ....` pull these together to make the function more readable

Reply:
done